modulemd/tests/test-modulemd-common.c

```diff
@@ -82,6 +82,84 @@ test_modulemd_load_file (void)
 }
 
 
+static void
+test_modulemd_load_string (void)
+{
+  const gchar *yaml_string = NULL;
+  g_autoptr (GError) error = NULL;
+  g_autoptr (ModulemdModuleIndex) idx = NULL;
+  g_autofree gchar *output = NULL;
+
+  /* This function is a wrapper around lower-level functions, so it should be
+   * okay to just test basic success and failure here.
+   */
+
+  /* Trivial modulemd */
+  g_clear_error (&error);
+  g_clear_object (&idx);
+  yaml_string =
+    "---\n"
+    "document: modulemd\n"
+    "version: 2\n"
+    "data:\n"
+    "  name: trivialname\n"
+    "  stream: trivialstream\n"
+    "  summary: Trivial Summary\n"
+    "  description: >-\n"
+    "    Trivial Description\n"
+    "  license:\n"
+    "    module: MIT\n"
+    "...\n";
+  idx = modulemd_load_string (yaml_string, &error);
+  g_assert_no_error (error);
+  g_assert_nonnull (idx);
+
+  /* Make sure loaded index dumps to string cleanly */
+  g_clear_error (&error);
+  g_clear_pointer (&output, g_free);
+  output = modulemd_module_index_dump_to_string (idx, &error);
+  g_assert_no_error (error);
+  g_assert_nonnull (output);
+
+
+  /* NULL string should raise an exception */
+  g_clear_error (&error);
+  g_clear_object (&idx);
+  modulemd_test_signal = 0;
+  signal (SIGTRAP, modulemd_test_signal_handler);
+  idx = modulemd_load_string (NULL, &error);
+  g_assert_cmpint (modulemd_test_signal, ==, SIGTRAP);
+  g_assert_null (idx);
+
+
+  // An empty string returns non-NULL but empty index. Is that right?
+  g_clear_error (&error);
+  g_clear_object (&idx);
+  yaml_string = "";
+  idx = modulemd_load_string (yaml_string, &error);
+  g_assert_no_error (error);
+  g_assert_nonnull (idx);
+
+  // Outputting the empty index returned above returns a validation error...
+  g_clear_error (&error);
+  g_clear_pointer (&output, g_free);
+  output = modulemd_module_index_dump_to_string (idx, &error);
+  g_assert_error (error, MODULEMD_ERROR, MMD_ERROR_VALIDATE);
+  g_assert_null (output);
+
+
+  /* Invalid YAML string */
+  g_clear_error (&error);
+  g_clear_object (&idx);
+  yaml_string = "Hello, World!\n";
+  idx = modulemd_load_string (yaml_string, &error);
+  // Should it actually return MMD_YAML_ERROR_UNPARSEABLE?
+  // Or should it just assert it's anything in the MODULEMD_YAML_ERROR domain?
```

Comment:
Probably we should clarify the definition of those two errors.

`MMD_YAML_ERROR_UNPARSEABLE` means that libyaml has thrown an error trying to interpret the data. For example, `modulemd_load_string ('[', &error);` would throw it because it encountered the start of an array but never the end of it.

It's basically the difference between "violates the YAML spec" and "violates the modulemd spec", if that helps.

Reply:
Got it. I'll try to add some clarifying text for those two errors in `modulemd/include/modulemd-2.0/modulemd-errors.h`.